### DISCUSSION TITLE
2.7.x: Problems consuming tycho built artifacts in non tycho builds

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-tycho/tycho#989] [Q&A]
Hi, using the xtext-maven-plugin with dependencies built by tycho 2.7.x we get following error message

```
Cannot access https://download.eclipse.org/releases/2022-06/202205271000/ with type p2 using the available connector factories: BasicRepositoryConnectorFactory: Cannot access https://download.eclipse.org/releases/2022-06/202205271000/ with type p2 using the available layout factories: Maven2RepositoryLayoutFactory: Unsupported repository layout p2
```

steps to reproduce
```
git clone
https://github.com/xtext/maven-xtext-example.git
cd maven-xtext-example
git checkout cd_tychoProblem
./build.sh
```

as a workaround it seems to work to switch back to tycho 2.6.0
or configure tycho also in the example-project/pom.xml

is this behaviour change intentional?

**Comment by laeubi** [comment]
> get following error message

is it a build error (thus aborting the build) or is it just a "warning" that this repository is not accessible?

> or configure tycho also in the example-project/pom.xml

You could try out tycho 3.0 and include the `org.eclipse.tycho:p2-maven-plugin` as it includes the p2 repository layout.

> as a workaround it seems to work to switch back to tycho 2.6.0

How could you "switch back" in a build not using tycho?

**Reply by cdietrich** [reply]
the build on the non tycho project fails with errrors. 
with switch back i mean: build the p2 plugins with tycho 2.6, then it does not care about p2 not avaiable in the non tycho project

**Reply by laeubi** [reply]
It might be that the plain maven project is confused about the new maven meta-data, so you might want to check the exact error and maybe disable the tycho consumer pom generation or further tweak it with the flatten-pom plugin.

**Reply by cdietrich** [reply]
what do you mean by exact error?

```
[ERROR] Failed to execute goal org.eclipse.xtext:xtext-maven-plugin:2.27.0:generate (default) on project example-project: Execution default of goal org.eclipse.xtext:xtext-maven-plugin:2.27.0:generate failed: Plugin org.eclipse.xtext:xtext-maven-plugin:2.27.0 or one of its dependencies could not be resolved: Could not transfer artifact org.osgi:org.osgi.service.prefs:eclipse-plugin:1.1.2 from/to eclipse (https://download.eclipse.org/releases/2022-06/202205271000/): Cannot access https://download.eclipse.org/releases/2022-06/202205271000/ with type p2 using the available connector factories: BasicRepositoryConnectorFactory: Cannot access https://download.eclipse.org/releases/2022-06/202205271000/ with type p2 using the available layout factories: Maven2RepositoryLayoutFactory: Unsupported repository layout p2 -> [Help 1]
```

how can i disable the tycho consumer pom generation

**Reply by laeubi** [reply]
> what do you mean by exact error?

In your initial question you claimed the error is a missing p2 repository but I more suspect its a missing artifact and thats was triggering the error output (that is then ignored by maven)

> Plugin org.eclipse.xtext:xtext-maven-plugin:2.27.0 or one of its dependencies could not be resolved: Could not transfer artifact org.osgi:org.osgi.service.prefs:eclipse-plugin:1.1.2

so this error indicated that the `xtext-maven-plugin` requires `org.osgi:org.osgi.service.prefs` here, are you using the latest Tycho 2.7.3 version? Why do you build a maven-plugin with tycho?

> how can i disable the tycho consumer pom generation

Set https://www.eclipse.org/tycho/sitedocs/tycho-packaging-plugin/update-consumer-pom-mojo.html#skipPomGeneration for the packaging plugin to `<skipPomGeneration>true</skipPomGeneration>`

**Reply by cdietrich** [reply]
skipping just leads to
```
[ERROR] Failed to execute goal org.eclipse.xtext:xtext-maven-plugin:2.27.0:generate (default) on project example-project: Execution default of goal org.eclipse.xtext:xtext-maven-plugin:2.27.0:generate failed: Plugin org.eclipse.xtext:xtext-maven-plugin:2.27.0 or one of its dependencies could not be resolved: Could not find artifact my.mavenized.herolanguage:my.mavenized.herolanguage:jar:1.0.0-SNAPSHOT -> [Help 1]
```
> xtext-maven-plugin requires org.osgi:org.osgi.service.prefs
<img width="1170" alt="Bildschirmfoto 2022-06-02 um 09 59 52" src="https://user-images.githubusercontent.com/238736/171583125-92e7a322-46e3-4881-b7e8-0b0150e83675.png">

i dont think so and even if it should be found right? why does it search it in p2 and not in maven central

https://github.com/xtext/maven-xtext-example/blob/aee677af8fb57acd17421cf1b472605e8dae168e/example-project/pom.xml#L95

and why is it working with tycho 2.6

a

**Reply by cdietrich** [reply]
i checked the pom generated
it looks like it has
```
    <dependency>
      <groupId>org.osgi</groupId>
      <artifactId>org.osgi.service.prefs</artifactId>
      <version>1.1.2</version>
      <type>eclipse-plugin</type>
      <scope>compile</scope>
      <optional>false</optional>
    </dependency>
```

**Reply by cdietrich** [reply]
and with 2.6.0 no such deps are generated

**Reply by laeubi** [reply]
> skipping just leads to

You should skip that for the xtext-maven plugin build, and then the pom should not contain any extra dependencies.

**Reply by cdietrich** [reply]
the xtext maven plugin is not but with tycho at all 
this p2 plugin is https://github.com/xtext/maven-xtext-example/blob/aee677af8fb57acd17421cf1b472605e8dae168e/example-project/pom.xml#L95
is added the skip to the build of exactly that

**Reply by laeubi** [reply]
> the xtext maven plugin is not but with tycho at all

The build-output claims differently:

> Plugin org.eclipse.xtext:xtext-maven-plugin:2.27.0 **or one of its dependencies** could not be resolved

so either the `xtext-maven-plugin` or any of its dependencies are build with tycho and include these extra requirements.

**Reply by cdietrich** [reply]
there is two maven builds.

- Tycho build
  - builds my.mavenized.herolanguage p2 plugin and others
  - tycho 2.7 will add add. deps to my.mavenized.herolanguage  generated pom
- Maven build
  - uses xtext-maven-plugin. adds my.mavenized.herolanguage as extra dependency
  
  depending on the tycho version in the first build, the 2nd build is working or not
  wil give the skip another try

**Reply by laeubi** [reply]
Okay, I see, yes then you need to skip the pom generation for the hero-language (why is it build with Tycho if it is meant to be used in maven?).

**Reply by cdietrich** [reply]
cause it is meant to be used for both
as e.g. a lot of eclipse platform  and emf is

**Reply by cdietrich** [reply]
ok ths one seems to work https://github.com/xtext/maven-xtext-example/pull/82/files
